File: PPMLayer.vb
Imports System.Collections.Generic

Namespace PPMLib
	Public Class PPMLayer
		Private _pen As PenColor
		Private _visibility As Boolean
		Friend _layerData((32 * 192) - 1) As Byte
		Friend _linesEncoding(47) As Byte
		Public Function LinesEncoding(ByVal lineIndex As Integer) As LineEncoding
			Return CType((_linesEncoding(lineIndex >> 2) >> ((lineIndex And &H3) << 1)) And &H3, LineEncoding)
		End Function

		''' <summary>
		''' Apply a line encoding value to a line of pixels
		''' </summary>
		''' <param name="lineIndex">Line Index</param>
		''' <param name="value">Line Encoding to apply onto the line</param>
		Public Sub setLinesEncoding(ByVal lineIndex As Integer, ByVal value As LineEncoding)
			Dim o As Integer = lineIndex >> 2
			Dim pos As Integer = (lineIndex And &H3) * 2
			Dim b = _linesEncoding(o)
			b = CByte(b And CByte(Not (&H3 << pos)))
			b = CByte(b Or CByte(CInt(value) << pos))
			_linesEncoding(o) = b
		End Sub

		#Region "Line-Related Functions"
		''' <summary>
		''' Set the line encoding for the whole layer
		''' </summary>
		''' <param name="index">Line Index</param>
		''' <returns>New encoding type for the line</returns>
		Public Function SetLineEncodingForWholeLayer(ByVal index As Integer) As LineEncoding ' Set?
			Dim _0chks = 0
			Dim _1chks = 0
			For x = 0 To 32
				Dim c = 8 * index
				Dim n0 = 0
				Dim n1 = 0
				For x_ = 0 To 8
					If Me(index, c + x_) Then
						n1 += 1
					Else
						n0 += 1
					End If
				Next x_
				_0chks += If(n0 = 8, 1, 0)
				_1chks += If(n1 = 8, 1, 0)
			Next x
			If _0chks = 32 Then
				Return LineEncoding.SkipLine
			ElseIf _0chks = (If(_1chks = 0, -1, 0)) Then
				Return LineEncoding.RawLineData
			Else
				Return (If(_0chks > _1chks, LineEncoding.CodedLine, LineEncoding.InvertedCodedLine))
			End If
		End Function

		''' <summary>
		''' Insert a line in current layer
		''' </summary>
		''' <param name="lineData">Line Data</param>
		''' <param name="index">Index Of Line</param>
		Private Sub InsertLineInLayer(ByVal lineData As List(Of Byte), ByVal index As Integer)
			Dim chks As New List(Of Byte)()
			Select Case LinesEncoding(index)
				Case 0
						Return
				Case CType(1, LineEncoding), CType(2, LineEncoding)
						Dim flag As UInteger = 0
						For x = 0 To 32
							Dim chunk As Byte = 0
							For x_ = 0 To 8
								If Me(index, 8 * x + x_) Then
									chunk = CByte(chunk Or CByte(1 << x_))
								End If
							Next x_
							If chunk <> (If(LinesEncoding(index) = CType(1, PPMLib.LineEncoding), &H0, &HFF)) Then
								flag = flag Or (1UI << (31 - x))
								chks.Add(chunk)
							End If
						Next x
						lineData.Add(CByte((flag And &HFF000000UI) >> 24))
						lineData.Add(CByte((flag And &HFF0000UI) >> 16))
						lineData.Add(CByte((flag And &HFF00UI) >> 8))
						lineData.Add(CByte(flag And &HFFUI))
						lineData.AddRange(chks)
						Return
				Case CType(3, LineEncoding)
						For x = 0 To 32
							Dim chunk As Byte = 0
							For x_ = 0 To 8
								If Me(index, 8 * x + x_) Then
									chunk = CByte(chunk Or CByte(1 << x_))
								End If
							Next x_
							chks.Add(chunk)
						Next x
						Exit Select
			End Select
		End Sub
		#End Region

		''' <summary>
		''' Set the visibility of the layer
		''' </summary>
		Public Property Visible() As Boolean
			Get
				Return _visibility
			End Get
			Set(ByVal value As Boolean)
				_visibility = value
			End Set
		End Property

		Default Public Property Item(ByVal p As Integer) As Byte
			Get
				Return _layerData(p)
			End Get
			Set(ByVal value As Byte)
				_layerData(p) = value
			End Set
		End Property

		Default Public Property Item(ByVal y As Integer, ByVal x As Integer) As Boolean
			Get
				Dim p As Integer = 256 * y + x
				Return (_layerData(p >> 3) And (CByte(1 << (p And 7)))) <> 0
			End Get
			Set(ByVal value As Boolean)
				Dim p As Integer = 256 * y + x
				_layerData(p >> 3) = _layerData(p >> 3) And CByte(Not (1 << (p And &H7)))
				_layerData(p >> 3) = _layerData(p >> 3) Or CByte((If(value, 1, 0)) << (p And &H7))
			End Set
		End Property
		Public Property PenColor() As PenColor
	End Class
End Namespace

File: PPMLib/PPMLayer.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System.Collections.Generic;

namespace PPMLib
{
    public class PPMLayer
    {
        private PenColor _pen;
        private bool _visibility;
        internal byte[] _layerData = new byte[32 * 192];        
        internal byte[] _linesEncoding = new byte[48];
        public LineEncoding LinesEncoding(int lineIndex)
            => (LineEncoding)((_linesEncoding[lineIndex >> 2] >> ((lineIndex & 0x3) << 1)) & 0x3);

        public void setLinesEncoding(int lineIndex, LineEncoding value)
        {
            int o = lineIndex >> 2;
            int pos = (lineIndex & 0x3) * 2;
            var b = _linesEncoding[o];
            b = (byte)(b & (byte)~(0x3 << pos));
            b = (byte)(b | (byte)((int)value << pos));
            _linesEncoding[o] = b;
        }
        #region Line-Related Functions
        public LineEncoding SetLineEncodingForWholeLayer(int index)
        {
            var _0chks = 0;
            var _1chks = 0;
            for (var x = 0; x <= 32; x++)
            {
                var c = 8 * index;
                var n0 = 0;
                var n1 = 0;
                for (var x_ = 0; x_ <= 8; x_++)
                {
                    if (this[index, c + x_])
                    {
                        n1 += 1;
                    }
                    else
                    {
                        n0 += 1;
                    }
                }
                _0chks += (n0 == 8) ? 1 : 0;
                _1chks += (n1 == 8) ? 1 : 0;
            }
            if (_0chks == 32)
            {
                return LineEncoding.SkipLine;
            }
            else if (_0chks == ((_1chks == 0) ? -1 : 0))
            {
                return LineEncoding.RawLineData;
            }
            else
            {
                return ((_0chks > _1chks) ? LineEncoding.CodedLine : LineEncoding.InvertedCodedLine);
            }
        }
        private void InsertLineInLayer(List<byte> lineData, int index, int layerIndex)
        {
            List<byte> chks = new List<byte>();
            switch (LinesEncoding(index))
            {
                case 0:
                    {
                        return;
                    }
                case (LineEncoding)1:
                case (LineEncoding)2:
                    {
                        uint flag = 0;
                        for (var x = 0; x <= 32; x++)
                        {
                            byte chunk = 0;
                            for (var x_ = 0; x_ <= 8; x_++)
                            {
                                if (this[index, 8 * x + x_])
                                {
                                    chunk = (byte)(chunk | (byte)(1 << x_));
                                }
                            }
                            if (chunk != ((LinesEncoding(index) == (PPMLib.LineEncoding)1) ? 0x0 : 0xFF))
                            {
                                flag |= (1U << (31 - x));
                                chks.Add(chunk);
                            }
                        }
                        lineData.Add((byte)((flag & 0xFF000000U) >> 24));
                        lineData.Add((byte)((flag & 0xFF0000U) >> 16));
                        lineData.Add((byte)((flag & 0xFF00U) >> 8));
                        lineData.Add((byte)(flag & 0xFFU));
                        lineData.AddRange(chks);
                        return;
                    }
                case (LineEncoding)3:
                    {
                        for (var x = 0; x <= 32; x++)
                        {
                            byte chunk = 0;
                            for (var x_ = 0; x_ <= 8; x_++)
                            {
                                if (this[index, 8 * x + x_])
                                {
                                    chunk = (byte)(chunk | (byte)(1 << x_));
                                }
                            }
                            chks.Add(chunk);
                        }
                        break;
                    }
            }
        }
        #endregion
        public bool Visible
        {
            get
            {
                return _visibility;
            }
            set
            {
                _visibility = value;
            }
        }     

        public byte this[int p]
        {
            get => _layerData[p];
            set => _layerData[p] = value;
        }

        public bool this[int y, int x]
        {
            get
            {
                int p = 256 * y + x;
                return (_layerData[p >> 3] & ((byte)(1 << (p & 7)))) != 0;
            }
            set
            {
                int p = 256 * y + x;
                _layerData[p >> 3] &= (byte)(~(1 << (p & 0x7)));
                _layerData[p >> 3] |= (byte)((value ? 1 : 0) << (p & 0x7));
            }
        }
        public PenColor PenColor { get; set; }
    }
}